ssh-server/main.go

```diff
@@ -113,38 +118,96 @@ func serviceChannel(newChannel ssh.NewChannel) error {
 	}
 	defer channel.Close()
 
+	allow := map[string]bool{"shell": true, "pty-req": true, "window-change": true}
+	resize := make(chan *pty.Winsize, 1)
+	defer close(resize)
+	done := make(chan error)
+	for {
+		select {
+		case err := <-done:
+			if err != nil {
+				return fmt.Errorf("running pty: %v", err)
+			}
+			return nil
+		case req, ok := <-requests:
+			if !ok {
+				return nil
+			}
+			if req.WantReply {
+				req.Reply(allow[req.Type], nil)
+			}
+			switch req.Type {
+			case "pty-req":
+				r := bytes.NewReader(req.Payload)
+				term, err := readString(r)
+				if err != nil {
+					return fmt.Errorf("parsing TERM environment variable value: %v", err)
+				}
+
+				if err := readWindowSize(r, resize); err != nil {
+					return fmt.Errorf("reading window size: %v", err)
+				}
+
+				go func() {
+					done <- runPTY(channel, term, resize)
+				}()
+			case "window-change":
+				if err := readWindowSize(bytes.NewReader(req.Payload), resize); err != nil {
+					return fmt.Errorf("reading window size: %v", err)
+					continue
+				}
+			}
+		}
+	}
+}
+
+func runPTY(channel ssh.Channel, term string, resize chan *pty.Winsize) error {
 	// Start the command with a pseudo-terminal.
-	shell, err := pty.Start(exec.Command("bash"))
+	cmd := exec.Command("bash")
+	cmd.Env = append(cmd.Env, fmt.Sprintf("TERM=%s", term))
+	shell, err := pty.Start(cmd)
 	if err != nil {
 		return fmt.Errorf("starting pty: %v", err)
 	}
 	defer shell.Close()
 
-	go func(in <-chan *ssh.Request) {
-		for req := range in {
-			// Tell the client we accept pty and shell commands
-			req.Reply(req.Type == "pty-req" || req.Type == "shell", nil)
-
-			if req.Type == "pty-req" {
-				skip := binary.BigEndian.Uint32(req.Payload)
-				resize(shell, req.Payload[4+skip:])
-			}
-			if req.Type == "window-change" {
-				resize(shell, req.Payload)
-			}
+	go func() {
+		for size := range resize {
+			pty.Setsize(shell, size)
 		}
-	}(requests)
+	}()
 
 	go io.Copy(shell, channel)
 	io.Copy(channel, shell)
+
+	status := struct{ Status uint32 }{}
+	if _, err := channel.SendRequest("exit-status", false, ssh.Marshal(&status)); err != nil {
+		return fmt.Errorf("sending exit status: %v", err)
+	}
 	return nil
 }
 
-func resize(f *os.File, payload []byte) {
+func readWindowSize(r io.Reader, resize chan *pty.Winsize) error {
```

Comment:
this function does more than reading - it applies the change too.  handleResizeRequest?

Reply:
The only extra work that it does is to write the window size to the output channel, the actual resize is done in runPty(). I renamed the channel parameter to say output and now I think it's clear enough that the function reads to a channel instead of returning the value. I think handleResizeRequest is not generic enough for both window-change and pty requests and it implies it does the actual resize

Reply:
As discussed, it's odd to have a read function not returning a value.  Let's have it return the value and you can have the caller send the value on the channel.